product_combination_exclude/models/product_template.py

```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Graeme Gellatly
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import itertools
+import psycopg2
+
+from odoo import api, models
+from odoo import tools
+from odoo.exceptions import except_orm
+
+
+class ProductTemplate(models.Model):
+
+    _inherit = 'product.template'
+
+    @api.multi
+    def create_variant_ids(self):
+        res = super(ProductTemplate, self).create_variant_ids()
+        exclusion_obj = self.env['product.attribute.exclude']
+        product_obj = self.env['product.product']
+        for template in self:
+            exclusion_values = exclusion_obj.search(
+                ['|',
+                 ('product_tmpl_ids', '=', template.id),
+                 ('product_tmpl_ids', 'in', [])]
+            )
+
+            excl_attr_val_ids = exclusion_values.read(['attribute_value_ids'])
+            excl_attr_val_ids = list(set(itertools.chain.from_iterable(
+                [x.get('attribute_value_ids', []) for x in excl_attr_val_ids]
+            )))
+
+            products = product_obj.search([
+                ('product_tmpl_id', '=', template.id),
+                ('attribute_value_ids', 'in', excl_attr_val_ids)])
+
+            variants_to_unlink = products.filtered(
+                lambda p: any([e.attribute_value_ids <= p.attribute_value_ids
+                               for e in exclusion_values]))
+
+            variants_to_deactivate = []
+            for variant in variants_to_unlink:
+                try:
+                    with self._cr.savepoint(), tools.mute_logger(
+                            'odoo.sql_db'):
+                        variant.unlink()
+                # We catch all kind of exception to be sure that the operation
+                # doesn't fail.
+                except (psycopg2.Error, except_orm):
+                    variants_to_deactivate.append(variant.id)
+                    pass
```

Comment:
@lasley I would wager I've spent far more time dealing with product variants than you have.  I've also spent far more time dealing with each iteration of odoo code in regards to this functionality.   

You can call it cut and paste if you like, however I have rewritten that upstream function to be literally orders of magnitude faster, and more modular.  One day if its ever accepted I would change this to match as well although after 3 years I hold little hope.  The whole Odoo way of doing it is very inefficient, maybe 20 hours for just 20,000 variants. 

In any case, specifically, this error SHOULD pass silently as it isn't really an error, just a shortcut to avoid manually searching all foreign key relations.  

Fundamentally it is just saying delete if we can, deactivate if we can't.  The failure is expected behaviour and should not be logged.